src/Nethermind/Nethermind.Evm/EvmStack.cs

```diff
@@ -241,56 +273,83 @@ public void PopUInt256(out UInt256 result)
-                u3 = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ulong>(ref bytes));
-                u2 = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, sizeof(ulong))));
-                u1 = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, 2 * sizeof(ulong))));
-                u0 = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, 3 * sizeof(ulong))));
+                Word data = Unsafe.ReadUnaligned<Word>(ref bytes);
+                Word shuffle = Vector256.Create(
+                    (byte)
+                    31, 30, 29, 28, 27, 26, 25, 24,
+                    23, 22, 21, 20, 19, 18, 17, 16,
+                    15, 14, 13, 12, 11, 10, 9, 8,
+                    7, 6, 5, 4, 3, 2, 1, 0);
+                Word convert = Avx2.Shuffle(data, shuffle);
+                Vector256<ulong> permute = Avx2.Permute4x64(Unsafe.As<Word, Vector256<ulong>>(ref convert), 0b_01_00_11_10);
+                result = Unsafe.As<Vector256<ulong>, UInt256>(ref permute);
             }
             else
             {
-                u3 = Unsafe.ReadUnaligned<ulong>(ref bytes);
-                u2 = Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, sizeof(ulong)));
-                u1 = Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, 2 * sizeof(ulong)));
-                u0 = Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, 3 * sizeof(ulong)));
+                ulong u3, u2, u1, u0;
+                if (BitConverter.IsLittleEndian)
+                {
+                    // Combine read and switch endianness to movbe reg, mem
+                    u3 = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ulong>(ref bytes));
+                    u2 = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, sizeof(ulong))));
+                    u1 = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, 2 * sizeof(ulong))));
+                    u0 = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, 3 * sizeof(ulong))));
+                }
+                else
+                {
+                    u3 = Unsafe.ReadUnaligned<ulong>(ref bytes);
+                    u2 = Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, sizeof(ulong)));
+                    u1 = Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, 2 * sizeof(ulong)));
+                    u0 = Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bytes, 3 * sizeof(ulong)));
+                }
+
+                result = new UInt256(u0, u1, u2, u3);
             }
-
-            result = new UInt256(u0, u1, u2, u3);
         }
 
         public bool PeekUInt256IsZero()
         {
-            int head = Head - 1;
-            if (head <= 0)
+            int head = Head;
+            if (head-- == 0)
```

Comment:
`head--` or `--head` ?

Reply:
Post for pop; though pre and post confuse me :)

Reply:
I think the new behaviour doesn't match the old one : [sharplabExperiment](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUBgRjz2AFMBnYAgCgIEoBuEi4dWx43ZymzD3MpTZ8mXMYKrUAtPVES2M/nlQAWAAQSaASyjA1ACVIBDOHTUBvPGutqdegBbG4agLwGnom7YBma6o5MpKVc3PgsrLxtUAgBOagASACIoUgB3NQA3UjByLQB7WDUjMABzCABbUl01EAtDEwBfRP4vBrw2sVUNFjY7dxMzS1wvPoDnN3rnYIJPGy1ffyc1AB5QwYjItWi4pLyAG2csnPzC4rLK6trzSaaWmw6GoA===)

Reply:
Yeah was bug before; would fail if top of the stack; now it matches all the other versions